tests/msc3083_test.go

```diff
@@ -163,3 +164,160 @@ func TestRestrictedRoomsRemoteJoin(t *testing.T) {
+				},
+			},
+		},
+	})
+	alice.SendEventSynced(t, space, b.Event{
+		Type:     "org.matrix.msc1772.space.child",
+		StateKey: &room,
+		Content: map[string]interface{}{
+			"via": []string{"hs1"},
+		},
+	})
+
+	t.Logf("Space: %s", space)
+	t.Logf("Room: %s", room)
+
+	// Create a second user on the same homeserver.
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+
+	// Querying the space returns only the space, as the room is restricted.
+	requestAndAssertSummary(t, bob, space, []interface{}{space})
+
+	// Join the space, and now the restricted room should appear.
+	bob.JoinRoom(t, space, []string{"hs1"})
+	requestAndAssertSummary(t, bob, space, []interface{}{space, room})
+}
+
+// Tests that MSC2946 works over federation for a restricted room.
+//
+// Create a space with a room in it that has join rules restricted to membership
+// in that space. The space and room are on different homeservers.
+//
+// The user should be unable to see the room in the spaces summary unless they
+// are a member of the space.
+func TestRestrictedRoomsSpacesSummaryFederation(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintFederationTwoLocalOneRemote)
+	defer deployment.Destroy(t)
+
+	// Create the rooms
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+	space := alice.CreateRoom(t, map[string]interface{}{
+		"preset": "public_chat",
+		"initial_state": []map[string]interface{}{
+			{
+				"type":      "m.room.history_visibility",
+				"state_key": "",
+				"content": map[string]string{
+					"history_visibility": "world_readable",
+				},
+			},
+		},
+	})
+
+	// The room is an unstable room version which supports the restricted join_rule
+	// and is created on hs2.
+	charlie := deployment.Client(t, "hs2", "@charlie:hs2")
+	room := charlie.CreateRoom(t, map[string]interface{}{
+		"preset":       "public_chat",
+		"name":         "Room",
+		"room_version": "org.matrix.msc3083",
+		"initial_state": []map[string]interface{}{
+			{
+				"type":      "m.room.join_rules",
+				"state_key": "",
+				"content": map[string]interface{}{
+					"join_rule": "restricted",
+					"allow": []map[string]interface{}{
+						{
+							"space": &space,
+							"via":   []string{"hs1"},
+						},
+					},
+				},
+			},
+		},
+	})
+
+	// create the link (this doesn't really make sense since how would alice know
+	// about the room? but it works for testing)
+	alice.SendEventSynced(t, space, b.Event{
+		Type:     spaceChildEventType,
+		StateKey: &room,
+		Content: map[string]interface{}{
+			"via": []string{"hs2"},
+		},
+	})
+
+	// The room appears for no one at first since hs2 doesn't know about who is in ss1.
+	requestAndAssertSummary(t, alice, space, []interface{}{space})
+	requestAndAssertSummary(t, bob, space, []interface{}{space})
+
+	// Charlie joins ss1 and now hs2 knows that alice is in it.
+	charlie.JoinRoom(t, space, []string{"hs1"})
+
+	// The restricted room should appear for alice (who is in the space).
```

Comment:
This is the case as hs1 has to ask hs2 if Alice can view the room, and hs2 will only answer positively once hs2 knows that Alice is in the space (once charlie:hs2 has joined it)?

It'd be useful to explain the dependency chain for Alice being able to see the room, as on first glance Alice is in the space, so you'd assume that she'd immediately be able to see the room.

The docstring for this function simplifies this to:

```golang
// The user should be unable to see the room in the spaces summary unless they
// are a member of the space.
```

but it'd be useful to mention how multiple homeservers play into it without diving into the MSC.

Reply:
@anoadragon453 I reworded some of these comments (and expanded the docstring). I hope it is clearer now, but please let me know if it can be improved!